File: DFS/51N皇后.cpp
#include <vector>
#include <string>
using namespace std;
class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<string>grid(n,string(n,'.'));
        dfs(grid,0);
        return ans;
    }
    void dfs(vector<string>grid,int now){
        if(now>=grid.size()){
            ans.push_back(grid);
            return;
        }
        int n=grid.size();
        for(int i=0;i<n;i++){
            grid[now][i]='Q';
            if(check(grid,now,i)){
                dfs(grid,now+1);
            }
            grid[now][i]='.';
        }
    }
    bool check(vector<string>&grid,int x,int y){
        for(int i=0;i<x;i++){
            if(grid[i][y]=='Q')return false;
        }
        for(int i=x-1,j=y-1;j>=0&&i>=0;i--,j--){
            if(grid[i][j]=='Q')return false;
        }
        for(int i=x-1,j=y+1;i>=0&&y<grid.size();i--,j++){
            if(grid[i][j]=='Q')return false;
        }
        return true;
    }
    vector<vector<string>>ans;
};

File: DFS/51N皇后.py
from typing import List
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        grid = [['.' for _ in range(n)] for _ in range(n)]
        ans=[]
        def feasible(n:int,now:int,line:int) -> bool:
            for i in range(0,now):
                if grid[i][line] == 'Q':
                    return False
            i = 1
            while now - i >=0 and line - i >=0:
                if grid[now - i][line - i] == 'Q':
                    return False
                i += 1
            i = 1
            while now - i >=0 and line + i <n:
                if grid[now - i][line + i] =='Q':
                    return False
                i += 1
            return True
        def Nqueens(n,now):
            if now == n:
                nonlocal ans
                ans.append([''.join(row) for row in grid])
                return
            for i in range(0,n):
                grid[now][i]='Q'
                if(feasible(n,now,i)):
                    Nqueens(n,now+1)
                grid[now][i]='.'
            return 
        Nqueens(n,0)
        return ans